### PR TITLE
Fix nested int caching for NJT inference tensor

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #132814
* #132414
* __->__ #133061
* #130292
* #131803
* #131916

Tries to solve https://fb.workplace.com/groups/296513499814673/permalink/465746352891386/

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/_subclasses/fake_tensor.py [hunk 1]
```diff
@@ -572,9 +583,10 @@ def __set__(self, obj: FakeTensor, value: Optional[torch.SymInt]) -> None:
             setattr(obj, self._memo(obj), None)
             setattr(obj, self._memo_vc(obj), None)
             setattr(obj, self._memo_epoch(obj), None)
-        elif not torch.is_inference_mode_enabled():
+        elif not torch.is_inference_mode_enabled() or self._opt_vc:
```
Comment (soulitzer) [inline torch/_subclasses/fake_tensor.py:586]: I wonder why we don't just check `obj.is_inference()` here

Comment (ezyang) [reply torch/_subclasses/fake_tensor.py:586]: Yeah, is inference seems better

**Comment by ezyang** [comment]
In general, I don't think you should need an opt_vc. Instead, you check if you have an inference tensor / are in inference mode before you do a vc test. If you are, you skip the vc test.

**Comment by soulitzer** [comment]
> In general, I don't think you should need an opt_vc. Instead, you check if you have an inference tensor / are in inference mode before you do a vc test. If you are, you skip the vc test.

Hmm I feel like I don't completely understand. Are you proposing that opt_vc always be True? In my head the point of the vc test was to guard against the case where you call nonzero on x mutate x, then call nonzero on x again. You want the result of the first and second nonzero to have different shapes.

**Comment by soulitzer** [comment]
Closing as these changes have been added to the reland of the previous PR:  https://github.com/pytorch/pytorch/pull/133196

**Pull request closed without merging**